core/core.go

```diff
@@ -109,7 +110,13 @@ func (c *Core) procfutureHeaders() {
 	headers := make([]*types.Header, 0, c.futureHeaders.Len())
 	for _, hash := range c.futureHeaders.Keys() {
 		if header, exist := c.futureHeaders.Peek(hash); exist {
-			headers = append(headers, header.(*types.Header))
+			// Remove from the futureHeader cache if the block is more than 30 secs old
+			// If it is remove from the futureHeader cache
+			if uint64(time.Now().Unix()) > header.(*types.Header).Time()+futureHeaderTimeout {
```

Comment:
I'm not sure if this will work during sync. Say a block was produced yesterday, and you are syncing today (24hrs later).

`block.Time() + 30s` will be way less than `time.Now()`

Reply:
Maybe keep account of when the block was added into the list in a separate cache ?

Reply:
The futureHeaders cache needs a time field (lru.Cache may already have this), and _that_ is what you should use to determine if the header has been in the cache too long

Reply:
it doesn't have a time field. i just checked

Reply:
ok, I'll take this one over so you can focus on petx-fetching